pytorch_translate/common_layers.py

```diff
@@ -49,10 +50,57 @@ def forward(self, decoder_state, source_hids, src_lengths):
             output, attn_scores = self.dot_attention(
                 decoder_state, source_hids, src_lengths
             )
+        elif self.attention_type == 'mean' or self.attention_type == 'max':
+            output, attn_scores = self.pooling(
+                self.attention_type,
+                source_hids,
+                src_lengths,
+            )
         else:
             raise ValueError(f"Attention type {self.attention_type} is not supported")
         return output, attn_scores
 
+    def pooling(self, pool_type, source_hids, src_lengths):
+        assert self.decoder_hidden_state_dim == self.encoder_output_dim
+        max_srclen = source_hids.size()[0]
+        assert max_srclen == src_lengths.data.max()
+        batch_size = source_hids.size()[1]
+        src_indices = torch.arange(
+            0,
+            max_srclen,
+        ).unsqueeze(0).type_as(src_lengths.data)
+        src_indices = src_indices.expand(batch_size, max_srclen)
+
+        # expand from shape (batch_size,) to (batch_size, max_srclen)
+        src_lengths = src_lengths.unsqueeze(dim=1).expand(
+            batch_size,
+            max_srclen,
+        )
+        src_mask = Variable(
+            (src_indices < src_lengths.data).double().type_as(
+                source_hids.data,
+            ),
+            requires_grad=False,
+        ).t().unsqueeze(2)
+        denom = src_lengths.t().unsqueeze(2).type_as(source_hids)
+        if pool_type == "mean":
+            masked_hiddens = source_hids * src_mask
+            context = (masked_hiddens / denom).sum(dim=0)
+        elif pool_type == "max":
+            masked_hiddens = source_hids - 10e6 * (1 - src_mask)
```

Comment:
How do you choose this constant?

Reply:
This serves as a replacement for -infinity.  Anything larger than typical hidden outputs would work.  The reason we did not use -infinity is because it causes NaN's downstream which can break the rest of the code (implementation details).